src/wp-includes/html-api/class-wp-css-complex-selector.php

```diff
@@ -0,0 +1,192 @@
+	 * the element at index 1 is the combinator string constant from this class,
+	 * e.g. `WP_CSS_Complex_Selector::COMBINATOR_CHILD`.
+	 *
+	 * In the example selector below, an element like `<el class="selected">` is selected iff:
+	 *   - it is a child of an `H1` element
+	 *   - *and* that `H1` element is a descendant of a `HEADING` element.
+	 *
+	 * The `H1` and `HEADING` parts of this selector are the "context selectors." Note that this
+	 * terminology is used for purposes of this class but does not correspond to language in the
+	 * CSS or selector specifications.
+	 *
+	 * @example
+	 *
+	 *     $context_selectors
+	 *     ┏━━━━━━┻━━━━┓
+	 *     .heading h1 > el.selected
+	 *
+	 * The example would have the following relative selectors:
+	 *
+	 * @example
+	 *
+	 *     array (
+	 *       array(
+	 *         WP_CSS_Type_Selector( 'ident' => 'h1' ),
+	 *         '>', // WP_CSS_Complex_Selector::COMBINATOR_CHILD
+	 *       ),
+	 *       array(
+	 *         new WP_CSS_Type_Selector( 'header' ),
+	 *         ' ', // WP_CSS_Complex_Selector::COMBINATOR_DESCENDANT
+	 *       ),
+	 *     )
+	 *
+	 * Note that the order of context selectors is reversed. This is to match the self selector
+	 * first and then match the context selectors beginning with the selector closest to the self
+	 * selector.
+	 *
+	 * @readonly
+	 * @var array{WP_CSS_Type_Selector, string}[]|null
+	 */
+	public $context_selectors;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param WP_CSS_Compound_Selector $self_selector The selector in the final position.
+	 * @param array{WP_CSS_Type_Selector, string}[]|null $selectors The context selectors.
+	 */
+	public function __construct(
+		WP_CSS_Compound_Selector $self_selector,
+		?array $context_selectors
+	) {
+		$this->self_selector     = $self_selector;
+		$this->context_selectors = $context_selectors;
+	}
+
+	/**
+	 * Determines if the processor's current position matches the selector.
+	 *
+	 * @param WP_HTML_Processor $processor The processor.
+	 * @return bool True if the processor's current position matches the selector.
+	 */
+	public function matches( WP_HTML_Processor $processor ): bool {
+		// First selector must match this location.
+		if ( ! $this->self_selector->matches( $processor ) ) {
+			return false;
+		}
+
+		if ( null === $this->context_selectors || array() === $this->context_selectors ) {
+			return true;
+		}
+
+		$breadcrumbs = array_slice( array_reverse( $processor->get_breadcrumbs() ), 1 );
+		return $this->explore_matches( $this->context_selectors, $breadcrumbs );
+	}
+
+	/**
+	 * Checks for matches by recursively comparing context selectors with breadcrumbs.
+	 *
+	 * @param array{WP_CSS_Type_Selector, string}[] $selectors Selectors to match.
+	 * @param string[] $breadcrumbs Breadcrumbs.
+	 * @return bool True if a match is found, otherwise false.
+	 */
+	private function explore_matches( array $selectors, array $breadcrumbs ): bool {
```

Comment:
This is a recursive function. It's bounded by the length of the breadcrumbs list and the depth of the complex selector, but it may be good to set some limits here.